Last pass over README before PR

richb-hanover opened the pull request:
Here's an updated README that I'm satisfied with. (That doesn't mean it's correct - only that I cannot find anything else to take out...)

Open Issues:

- This first one might be controversial - I made the change to make it simpler to explain and use (I hope). That is, I rewrote the `latest` definition to use the most recent STABLE version. Mostly so that if a newcomer requests `jacobalberty/unifi` they get a stable version not a release candidate
- Consequently, I also changed the README to say that people who _do_ want a release candidate should look for `:development`
- I listed all the tag versions in a single table. I don't know how to know if they're all available...
- Have I described "override host IP" correctly?
- I tweaked up the formatting of Volumes and Environment Variables. Did I get it right?
- It's hard to know what to recommend for Windows users. The mongo/fsync() hassle is confusing. Could we simply recommend people use `docker-compose`? How would they supply options?
- I marked a couple sections at the end, asking if they are still relevant. (They seem to talk about much older versions of Docker and/or Unifi) Perhaps they should be updated or removed: I don't know enough about them to say for sure.

That's all the intentional changes I can remember. Please review this carefully - I'm kind of a newcomer myself to Docker and Unifi. 

All comments welcomed. Thanks.

richb-hanover commented:
> ... so lets set an arbitrary deadline deadline of say Wednesday the 13th.

Excellent. Please review carefully. I asserted a "bunch of stuff". Much of it is likely to be true ... :-) Thanks!

jacobalberty commented on README.md:
```diff
@@ -70,108 +88,118 @@ reusing the configuration from the local disk. The process is:
+_[**Note:** Docker's rule is to retrieve `:latest` if no tag is specified.
+Newcomers should not be using an RC version,
+so I suggest that we tag the most recent stable version with `:latest`.
+Someone who wants the release candidate can add the `:development` tag -richb]_
```
The way I handle this is `:latest` only contains the latest stable,  i have rcs under the `:rc` tag so its safe to use `:latest` for if you trust ubiquiti's definition of stable, though with unifi it is probably actually better to use a specific version tag like `:v7.1` for anything beyond home personal use

richb-hanover replied:
I agree. Perhaps we should only have `:latest` and `:7.1.66` tags (that is, the full version number, not `v7` nor `v7.1`)

richb-hanover replied:
... and to follow along with your thought, we should tell people to use the `:rc` tag if they want an RC. (Not my newly-invented `:development` tag...

jacobalberty commented:
I left a comment on one of your questions, I'm not sure on the windows one, it makes sense that it would work though.

richb-hanover commented:
I re-ordered the info:

- I moved the stuff I consider "important for newcomers" close the top
- Added a "Current Information" section. This would be a good place to put warnings about CVEs, etc.
- Moved up Volumes, Environment Variables, and Exposed Ports sections
- Moved discussion of Windows to the end.
- I boldly-asserted that Windows people should use docker-compose

Outstanding questions:
- Should we refer to beta branch in the Tag table? Or is it somehow different from the other tags?
- Multi-arch - is this still relevant? Or have all the tags listed in the table incorporated the multi-arch info?
- Multi-process containers - does docker-compose make this discussion obsolete?
- I don't understand how the certificate scripts work. If we want to expand on this, someone else will need to write this.
- Has anyone tested docker-compose and/or Windows recently?

Thanks!

jacobalberty commented:
I will be looking at this later today after work to see if its ready to merge or if we need to hold btw.

jacobalberty commented:
I'm going to have to push back reviewing and merging this until next week probably Monday or Tuesday.

richb-hanover commented:
OK. Thanks.

tiesmaster commented:
Hi guys, I've been a bit offline, so I missed this PR, and all the updates on it. Sorry for that, I can take a look at this tomorrow.

tiesmaster commented on README.md:
```diff
@@ -2,176 +2,178 @@
    -v ~/unifi:/unifi \
    --user unifi \
-   jacobalberty/unifi:v7
+   --name unifi \
```
👍

tiesmaster reviewed (commented):
@richb-hanover Very nice! I see you've spend a lot of time improving the docs. Good stuff!

I've briefly looked over the changes, but they're too much for me to do a proper pass on, so I've only looked at the `Override "Inform Host" IP` part, as that's were it all got started for me (on that see below), and I'm answering a couple questions from your original PR description, to help clarify some confusion:

> Have I described "override host IP" correctly?

No, I think this can still be improved a little bit better. See my feedback in the review.

> It's hard to know what to recommend for Windows users. The mongo/fsync() hassle is confusing. Could we simply recommend people use docker-compose? How would they supply options?

Actually, that doesn't make any sense. You're mixing up 2 different things here:

 * running Docker on a certain host system, such as Windows
 * making use of a container orchestrator, such as Docker Compose

Running MongoDB in Docker *with a bind mount* apparently doesn't work (haven't checked this myself, and the issue seems very old, Docker on Windows has switched to WSL in the mean time, and perhaps that actually does work), when running on Windows.

If you're using Docker Compose on Windows, then that's still going to be Docker Desktop underneath, and you'll suffer from the same issue. I'd recommend to make use of volume mounts, as indicated in the original `README.md`.

---

@richb-hanover @jacobalberty I hope you'll find my feedback useful, let me know if anything is unclear, or you need clarification otherwise.

tiesmaster commented:
> haven't checked this myself, and the issue seems very old, Docker on Windows has switched to WSL in the mean time, and perhaps that actually does work

I briefly looked into this, and this issue _seems_ to be solved. [WSL supports](https://docs.microsoft.com/en-us/windows/wsl/release-notes#wsl-8) `fsync` these days, and when running the latest `mongo` container, I don't _see_ the error floating by... So maybe, you just want to drop that from the `README.md`.

---

This is how I investigated:
I'm starting a new Docker container, like this (I'm using the bash shell via Git for Windows, so I need to [stop MSYS from "mangling the path"](https://stackoverflow.com/questions/7250130/how-to-stop-mingw-and-msys-from-mangling-path-names-given-at-the-command-line) I pass to Docker):

```
$ MSYS_NO_PATHCONV=1 docker run --rm -it -v c:/ties/tmp/mongo/test-data:/data/db -p 27017 mongo
```

which outputs (I don't see any errors as in [the referenced issue](https://github.com/docker/for-win/issues/138)):
![image](https://user-images.githubusercontent.com/460904/179292328-6f83c298-69fa-4645-8555-43a7a2286eea.png)

the bind mount is working, I see files of the datastore coming:
![image](https://user-images.githubusercontent.com/460904/179292480-a76f05c8-20ff-48b5-86bb-63360022fa05.png)

richb-hanover commented:
@tiesmaster Thanks for checking into these things that I don't know quite so much about.

1. Good catch on the path to the **Override Inform Host** option
2. Thanks for investigating Windows.
  * Do you think it's safe/appropriate simply to use the same instructions for Windows and other OS's?
  *  Do we need to recommend a certain version of Docker Desktop for Windows that uses WSL (and that is safe)?
3. Is there any reason to keep around the `docker-compose` discussion? I know a lot of hard work went into creating the `docker-compose.yml` file. It's very cool, but...
  * Does the `docker-compose` provide any benefits over the all-in-one container?
  * I made up a lot of baloney about `docker-compose` being a good solution for Windows. We should prune that out.
  * But that entire "Running Mongo in a separate container" section still makes the README longer, could confuse newcomers, and might not be a help to others. Should we just remove it? (Or does it provide value that I don't understand.)
4. Should `beta` be listed in the **Tags** table? (It could continue to have its own section farther down...)
5. What about the **Multiarch** and **Multi-process containers** sections? Is any of that information still relevant?

I think that's my list of outstanding questions. I know @jacobalberty won't have a chance to look at this 'til after the weekend.

tiesmaster commented:
> @tiesmaster Thanks for checking into these things that I don't know quite so much about.
> 
>     1. Good catch on the path to the **Override Inform Host** option

:+1:

> 
>     2. Thanks for investigating Windows.
> 
> 
>     * Do you think it's safe/appropriate simply to use the same instructions for Windows and other OS's?

Yes, indeed.

> 
>     * Do we need to recommend a certain version of Docker Desktop for Windows that uses WSL (and that is safe)?

Just the latest.

> 
> 
>     3. Is there any reason to keep around the `docker-compose` discussion? I know a lot of hard work went into creating the `docker-compose.yml` file. It's very cool, but...
> 
> 
>     * Does the `docker-compose` provide any benefits over the all-in-one container?

I don't think so. We use Docker Compose at work as orchestrator for development. That's a setup with multiple microservices, and dependencies, where it does make sense. And DC is a lot simpler, in its day to day usage, than something like Kubernetes. However, for this container, where you have a "all-in-one" container, I don't see that much value.

---

That's my $0.02, good luck finishing off this PR. I'll be on holiday for the upcoming 3 weeks, so don't expect any reply from me before then.

richb-hanover commented:
@tiesmaster Thanks for the thoughts. I made adjustments to the documentation to link to the Microsoft documentation for Docker on Windows. Also minor tweaks to wording.

jacobalberty commented on README.md:
```diff
@@ -439,6 +473,25 @@ If your certificate or private key have different names, you can set the environ
 
 For letsencrypt certs, we'll autodetect that and add the needed Identrust X3 CA Cert automatically. In case your letsencrypt cert is already the chained certificate, you can set the `CERT_IS_CHAIN` environment variable to `true`, e.g. `CERT_IS_CHAIN=true`. This option also works together with a custom `CERTNAME`.
 
+~~### multiarch~~
```
This section is relevant with the current tags, and specifically the stuff about `armhf` going away needs notice. I suspect a  new release of unifi will mandate `armhf` support being  removed within the next year.

richb-hanover replied:
I will move it back below the Tags section

jacobalberty commented on README.md:
```diff
@@ -439,6 +473,25 @@ If your certificate or private key have different names, you can set the environ
+
+~~## Multi-process container~~
+
+~~_[Is this section still true? Relevant? Does the `docker-compose`section supercede this?]_~~
```
Multi-process is the default for running docker without compose. There is no real need for the documentation on its behavior but outside of the included docker-compose.yml this behavior is still correct.

richb-hanover replied:
Good point. I have moved it to the Side Projects document

jacobalberty commented on README.md:
```diff
@@ -204,228 +215,251 @@ It is possible to configure the `macvlan` driver to bridge your container to the
+option on the `docker run...` to bind to whatever port you wish.
 
-Default `1024M`
+~~## Unifi-in-Docker on Windows~~
```
@tiesmaster says it works on windows without issue so we can probably drop this windows section. If someone reports its still an issue I can always add this section back again

richb-hanover replied:
Agree

jacobalberty commented on README.md:
````diff
@@ -204,228 +215,251 @@ It is possible to configure the `macvlan` driver to bridge your container to the
+```
 
-* 8443/tcp - Web interface + API
+## Beta Users
````
It might be worth it to remove the beta section for now, beta is somewhat broken unless run as root right now and needs some work to fix that.

richb-hanover replied:
OK

jacobalberty reviewed (commented):
I've added my notes on it the parts with questions.

Keeping multi-arch documentation is the only part I think is important to keep around.

Removing multi-process, windows, and docker-compose I'm game for stripping out, if there's confusion from the removal they can be added back in without issue.

Removing beta for now is probably a good idea as the beta tag only works when run as root right now, it need some updates to handle permissions properly with the deprecation of the RUNAS_UID0=false stuff.

richb-hanover commented:
This is great feedback. I will make those edits in the README.md, and also create an Additional-Info.md that preserves some of the stuff we are cutting out (beta, docker-compose) in case we need to refer someone to that info.

richb-hanover commented:
@jacobalberty I think I'm done. The two files - README and Side Projects are in the PR.

Unifi-in-Docker is ticking away on the Raspberry Pi in my office, keeping an eye on the 21 APs that I have sprinkled around [Loch Lyme Lodge](https://lochlymelodge.com). I haven't worried about it one little bit.

Thanks again for this great project.

richb-hanover commented:
Thank you!
